cpp/src/arrow/util/rle_encoding_internal.h

```diff
@@ -299,385 +552,988 @@ class RleEncoder {
+    -> values_count_type {
+  using ControlFlow = RleBitPackedParser::ControlFlow;
+
+  ARROW_DCHECK_GT(batch_size, 0);
+
+  auto batch = internal::BatchCounter::FromBatchSizeAndNulls(batch_size, null_count);
+
+  if (ARROW_PREDICT_FALSE(batch.IsFullyNull())) {
+    converter.WriteZero(out, out + batch.NullRemaining());
+    return batch.NullRemaining();
+  }
+
+  arrow::internal::BitRunReader validity_reader(validity_bits, validity_bits_offset,
+                                                /*length=*/batch.TotalRemaining());
+  arrow::internal::BitRun validity_run = validity_reader.NextRun();
+
+  auto const check_and_handle_fully_null_remaining = [&]() {
+    if (batch.IsFullyNull()) {
+      ARROW_DCHECK(validity_run.length == 0 || !validity_run.set);
+      ARROW_DCHECK_GE(validity_run.length, batch.NullRemaining());
+
+      converter.WriteZero(out, out + batch.NullRemaining());
+      out += batch.NullRemaining();
+      batch.AccrueReadNulls(batch.NullRemaining());
+    }
+  };
+
+  // Remaining from a previous call that would have left some unread data from a run.
+  if (ARROW_PREDICT_FALSE(RunRemaining() > 0)) {
+    auto const [values_read, null_read] =
+        RunGetSpaced(converter, out, batch.TotalRemaining(), batch.NullRemaining(),
+                     validity_reader, validity_run, decoder_);
+
+    batch.AccrueReadNulls(null_read);
+    batch.AccrueReadValues(values_read);
+    out += values_read + null_read;
+
+    // Either we fulfilled all the batch values to be read
+    if (ARROW_PREDICT_FALSE(batch.ValuesRemaining() == 0)) {
+      // There may be remaining null if they are not greedily filled
+      check_and_handle_fully_null_remaining();
+      return batch.TotalRead();
+    }
+
+    /// We finished the remaining run
+    ARROW_DCHECK(RunRemaining() == 0);
+  }
+
+  auto handler = internal::LambdaHandler{
+      [&](auto run) {
+        internal::decoder_for_t<value_type, decltype(run)> decoder(run);
+
+        auto const [values_read, null_read] = internal::RunGetSpaced(
+            converter, out, batch.TotalRemaining(), batch.NullRemaining(),
+            validity_reader, validity_run, decoder);
+
+        batch.AccrueReadNulls(null_read);
+        batch.AccrueReadValues(values_read);
+        out += values_read + null_read;
+
+        // Stop reading and store remaining decoder
+        if (ARROW_PREDICT_FALSE(values_read == 0 || batch.ValuesRemaining() == 0)) {
+          decoder_ = std::move(decoder);
+          return ControlFlow::Break;
+        }
+
+        return ControlFlow::Continue;
+      },
+  };
+
+  parser_.Parse(handler);
+
+  // There may be remaining null if they are not greedily filled by either decoder calls
+  check_and_handle_fully_null_remaining();
+
+  ARROW_DCHECK(batch.IsDone() || Exhausted());
+  // batch.Done() => batch.NullRemaining() == 0
+  ARROW_DCHECK(!batch.IsDone() || (batch.NullRemaining() == 0));
```

Comment:
If it's by definition of BatchCounter semantics, do we need to check for it here?